### DISCUSSION TITLE
Can this run on espruino js embeded devices?

### DISCUSSION THREAD

**Discussion by adminy** [mfucci/node-matter#51] [Q&A] [answered]
Since this is just using pure node libraries I don't see why it shouldn't be possible to run this on the espruino capable devices.

I have a few questions with this around topology and what the networking is in this.

Is it possible to run this as a border router?
io thread device with bluetooth le?

or is this just capable of orchestrating border routers?

can you explain what the hierarchy is here 🙏🏻

Can a device be a border router and a switch at the same time?

What other types of devices exist besides a basic switch?

Thank you

**Comment by mfucci** [comment, marked as answer]
Hi Adminy!

Thanks for the questions, answers below!

**Espruino**: yes, this is pure node / js code but I don't think it will run on espruino easily.
The code does use node.js specific API, for crypto / networking / data buffer.
With some work, it should be possible to make it browser-compatible, since similar API do exist on the browser.
By looking at Espruino API, as expected, the main API missing are for crypto:

- Spake2p crypto required crypto API that didn't exist in node.js, so I add to find pure js alternatives and this should work for espruino
- look at Crypto.ts, if you are able to implement all those function on espruino, then you should be good to go! (careful though, some API are only supported for some specific boards).
So if you are ready to take this challenge, start with an espruino version of Crypto.ts which might be the blocking point...

**Border router / network hierarchy**: you mean border router as in thread border router?
Matter protocol is the application on top, so is mostly agnostic of which underlying network system is used.
Only during commissioning (adding a device in the Matter mesh), there are network-specific pieces to discover the device and provide it with credentials to join a local network.
So node-matter goal is to act as a Matter device that can be controlled by a standard Matter controller, or at the opposite act as a Matter controller to control standard Matter devices.

For instance, you can build with it your smart home brain that will orchestrate all your automation systems.
I am currently implementing the API to do this, here is what I have in mind:
```
import { MatterController } from "./matter/MatterController";

const controller = new MatterController();
const temperatureSensor = controller.getNode(1).getCluster(TemperatureSensorCluster);
const acSwitch = controller.getNode(2).getCluster(OnOffCluster);

acSwitch.subscribeOn(on => {
    if (on) {
        console.log("AC has been turned on");
    } else {
        console.log("AC has been turned off");
    }
});

temperatureSensor.subscribeTemperature(temperature => {
    console.log(`Temperature is: ${temperature}`);
    const acOn = acSwitch.getOn();
    if (!acOn && temperature > Fahrenheit(80)) acSwitch.on();
    if (acOn && temperature < Fahrenheit(75)) acSwitch.off();
}, { interval: Minutes(5) });
```
You can have other services running on the same device so it can act as a thread border router / switch, but this is indenpendant.

At the opposite, you can use node-matter to access non-Matter device over Matter using a standard Matter controller (Apple / Google will provide one soon on Android / iOs).
For instance:
```
matter -on "echo 255 > /sys/class/leds/led1/brightness" -off "echo 0 > /sys/class/leds/led1/brightness"
```
This allow to control a raspberry Pi LED over Matter.

**What other types of devices exist besides a basic switch?**: you mean what other types of Matter devices exist?
You can find the official list of clusters [here](https://github.com/project-chip/connectedhomeip/tree/c438b8945e26a84f68ba3608de202e4b939a9080/src/app/clusters).
node-matter only implements the OnOff cluster as now, but adding the support for more cluster is straight forward.

**Reply by adminy** [reply]
The thing I'm afraid about the most is that I won't be able to fit in all the crypto scripts in the 1MB on the dongles with espruino. I might have to connect an external SPI flash to try and fill it all up. Good news is that the NRF52840 chip is capable of doing some of the functions out of the box. But the spi flash is undesireable. I'll see maybe I can minimise it before flashing it.

- Yes I meant thread border router.

Thanks for the list. Now I have more references that I needed, thanks.


for the API does it have to be specific about the node you need in the controller? can it be a scan query or a find for a given type(s) of node?

Something like:

```javascript
const temperatureSensor = matterController.scan({ types: TemperatureSensor })
// or
const temperatureSensor = matterController.nodes.find(cluster => cluster.type === TemperatureSensor)

```

**Reply by mfucci** [reply]
I think 1M should be plenty: not minified and including the client and the server, node-matter only takes 614k and this does include a library to do elliptic curve crypto.

APIs: I will start a new thread to discuss them. I will start by implementing the basic one though...

**Reply by mfucci** [reply]
Moved the API discussion to https://github.com/mfucci/node-matter/discussions/64

**Reply by mfucci** [reply]
I created a separate thread to follow support for Espruino effort

**Reply by adminy** [reply]
thank you

**Comment by Apollon77** [comment]
@mfucci For your proposal for a "MatterController API": As basic idea this is a good idea. In fact we will need some  ore methods on the controller to allow a "programmatic" understanding of the structure with multiple Bridged nodes, potentially used Labels and such stuff. In my eyes exactly knowing the details of the discovered matter devices is a rare case. In most cases you just use it and then need to be able to work with what you get ...

**Reply by mfucci** [reply]
Thanks! Moved the controller API discussion to https://github.com/mfucci/node-matter/discussions/64
